### PR TITLE
[NDRS-433] - Consensus using Timestamp::now() for timestamps

### PR DESCRIPTION
**Pull request by xcthulhu**
Highway now uses a timestamp set to `Timestamp::now()` where appropriate.

Exceptions to this are:

  1. Activity scheduled by the reactor (use the scheduled timestamp instead)
  2. Block proposals (use the block timestamp here)

This code assumes that messages with timestamps in the future have been rescheduled and all messages being processed by highway are in the past.

I have made a follow-up ticket to tackle rescheduling future messages here: https://casperlabs.atlassian.net/browse/NDRS-499

### CODE REVIEW

File: node/src/components/consensus/highway_core/highway_testing.rs
Comment (xcthulhu) [inline node/src/components/consensus/highway_core/highway_testing.rs:551]: @afck This should just be the heuristic logic that was there before we put in `Timestamp::now()`

File: node/src/components/consensus/highway_core/highway.rs
Comment (xcthulhu) [inline node/src/components/consensus/highway_core/highway.rs:265]: @afck When proposing a block, we should just use its timestamp, which will be a little older than `Timestamp::now()`...?

File: node/src/components/consensus/protocols/highway.rs
Comment (xcthulhu) [inline node/src/components/consensus/protocols/highway.rs:239]: @afck https://casperlabs.atlassian.net/browse/NDRS-499

File: node/src/components/consensus/highway_core/highway.rs
Comment (xcthulhu) [inline node/src/components/consensus/highway_core/highway.rs:249]: @afck Here in `handle_timer` we just use the timer's `timestamp`, and assume its ~ `Timestamp::now()`

File: node/src/components/consensus/highway_core/highway.rs
Comment (afck) [reply node/src/components/consensus/highway_core/highway.rs:265]: Yes, see comment above. What happens is basically this:
* We realize it's our turn to propose a block in millisecond 64, so we set a timer.
* The timer for timestamp 64 fires, and we request deploys for the new block from the deploy buffer (with 64 in the block context).
* The deploy buffer responds and we finally end up here, and can propose the new block. But we still have to use timestamp 64.

File: node/src/components/consensus/highway_core/highway_testing.rs
Comment (afck) [reply node/src/components/consensus/highway_core/highway_testing.rs:551]: Actually, at the call site of `process_message`, we still know the (simulated) `delivery_time` and could pass it in. Then pass it into `add_vertex` and we should have the "correct" timestamp here.

File: node/src/components/consensus/protocols/highway.rs
Comment (afck) [inline node/src/components/consensus/protocols/highway.rs:156]: As a next step (not necessarily this PR), this should be passed in here (and in `resolve_validity`), too, so that the `now()` call is made in `era_supervisor.rs` instead.

**Review by afck — COMMENTED** [review]
Looks good to me, but we could fix the `highway_testing.rs` `TODO` before merging.

File: node/src/components/consensus/highway_core/active_validator.rs
Comment (goral09) [inline node/src/components/consensus/highway_core/active_validator.rs:226]: Maybe a warning is enough for this? `timestamp < earliest_vote_time` can happen due to network issues, sender delays etc.

File: node/src/components/consensus/protocols/highway.rs
Comment (goral09) [reply node/src/components/consensus/protocols/highway.rs:156]: Shouldn't the timestamp for the vertex be chosen earlier? If it's "our vertex" it should come from the `BlockContext` and for an incoming vertex it should be part of it.

Comment (afck) [reply node/src/components/consensus/protocols/highway.rs:156]: Why? It doesn't make a difference whether we call `now()` here, or pass in `now()` as an argument.

(The timestamp we want to pass into `Highway` here is not the vertex', but _our_ current timestamp. E.g. if we received a proposal and are now making a confirmation, the confirmation should contain our own current time as its timestamp, not the proposal's.)

File: node/src/components/consensus/highway_core/active_validator.rs
Comment (afck) [reply node/src/components/consensus/highway_core/active_validator.rs:226]: I don't think it can happen directly due to network issues. `earliest_vote_time` is the timestamp of our own latest message, so this can only happen if a timer event got delayed so much that in the meantime we've already created a vertex with a later timestamp.

Comment (goral09) [reply node/src/components/consensus/highway_core/active_validator.rs:226]: But `timestamp` is a timestamp from the incoming message, right?

Comment (afck) [reply node/src/components/consensus/highway_core/active_validator.rs:226]: No, it used to be that, as a workaround. But with this PR, it will be the current timestamp, and thus the timestamp we would use for our confirmation vote.

Comment (afck) [reply node/src/components/consensus/highway_core/active_validator.rs:226]: (We access `vote.timestamp` below.)

Comment (goral09) [reply node/src/components/consensus/highway_core/active_validator.rs:226]: Thanks. I got lost a bit. How about renaming `timestamp` argument to `now`? Looks like this pattern is employed in other places.

Comment (afck) [reply node/src/components/consensus/highway_core/active_validator.rs:226]: I remember we talked about that last week. I forgot why we ended up with `timestamp` again. Sounds like it should be `now`, yes.

File: node/src/components/consensus/highway_core/active_validator.rs
Comment (xcthulhu) [reply node/src/components/consensus/highway_core/active_validator.rs]: Reverted.

File: node/src/components/consensus/highway_core/highway.rs
Comment (xcthulhu) [reply node/src/components/consensus/highway_core/highway.rs:249]: I have added a comment explaining this.

File: node/src/components/consensus/highway_core/highway_testing.rs
Comment (xcthulhu) [reply node/src/components/consensus/highway_core/highway_testing.rs:551]: Done.

**Comment by xcthulhu** [comment]
@afck I've tackled all of the TODOs, is there anything else or is this gtg?

File: node/src/components/consensus/highway_core/active_validator.rs
Comment (afck) [inline node/src/components/consensus/highway_core/active_validator.rs:129]: Should we call this `now`, too? (Also in `should_send_confirmation`.)

**Review by afck — APPROVED** [review]
Looks good to me, but let's also wait for @goral09's approval.

File: node/src/components/consensus/highway_core/active_validator.rs
Comment (xcthulhu) [reply node/src/components/consensus/highway_core/active_validator.rs]: Fixed in https://github.com/CasperLabs/casper-node/pull/354/commits/cdd18db9cea6cc1df8ea8cff9b2245fbd4637898

**Review by goral09 — APPROVED** [review]

**Comment by xcthulhu** [comment]
borse merge

**Comment by goral09** [comment]
@xcthulhu Bors failed b/c commits are not signed.

**Comment by xcthulhu** [comment]
bors merge

**Comment by xcthulhu** [comment]
bors merge

**Comment by xcthulhu** [comment]
bors merge

**Pull request merged**
